[Frontend] Some pipeline refactoring

hamishknight opened the pull request:
Move mangling stats, dependency verification, and API dumping to `performEndOfPipeline`. In the latter two cases this allows us to free the ASTContext before LLVM.

In addition, internalize the creation of the intermodule dependency tracker such that it's only set by `CompilerInstance::setup`.

hamishknight commented:
@swift-ci please test

CodaFi commented on lib/Frontend/Frontend.cpp:
```diff
@@ -288,9 +288,25 @@ void CompilerInstance::setupDiagnosticVerifierIfNeeded() {
+    // them directly; in particular, -track-system-dependencies affects how
+    // module interfaces get loaded, and so we need to be consistently tracking
+    // system dependencies throughout the compiler.
+    createDependencyTracker(opts.TrackSystemDeps);
```
It should be possible to internalize this API completely now. All the remaining callers will invoke `setup`. I think code completion might be a little tricky, but there's probably a quick fix there too.

hamishknight replied:
Ah, I didn't realise `createDependencyTracker` was being called elsewhere – let's expand the `FrontendOptions` to allow the dependency tracker to be enabled for non-system dependencies too. How does this look?

hamishknight committed:
[Frontend] Internalize createDependencyTracker

Expand the FrontendOptions to allow the enabling
of the dependency tracker for non-system
dependencies, and switch the previous clients of
`createDependencyTracker` over to using this
option. This ensures that the dependency tracker
is now set only during `CompilerInstance::setup`.

hamishknight committed:
[Frontend] Factor out setupDependencyTrackerIfNeeded

And call from `CompilerInstance::setup`.

hamishknight committed:
[Frontend] Move mangling stats to end of pipeline

hamishknight committed:
[Frontend] NFC: Move dumpAPI up a bit

hamishknight committed:
[Frontend] Move dependency verifier to end of pipeline

Allow the verification of dependencies before LLVM,
allowing the freeing of the ASTContext.

hamishknight committed:
[Frontend] dumpAPI at the end of the pipeline

This allows us to remove another condition that
was previously preventing the ASTContext from being
freed before LLVM.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/1392

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/1392

@swift-ci please test Linux

hamishknight commented:
https://github.com/apple/llvm-project/pull/1392

@swift-ci please test macOS

CodaFi commented on lib/Frontend/Frontend.cpp:
```diff
@@ -288,9 +288,36 @@ void CompilerInstance::setupDiagnosticVerifierIfNeeded() {
+  // interfaces get loaded, and so we need to be consistently tracking system
+  // dependencies throughout the compiler.
+  auto collectionMode = opts.IntermoduleDependencyTracking;
+  if (!collectionMode) {
```
Let's make this a `switch`
